Disabled field in compute host

AnishReddyRavula opened the pull request:
- Add disabled field in ComputeHost model
- - This can be used by admins to specify that a host is disabled for maintainance and that marks that blazar monitor should not touch or alter this host's health while it is disabled (True). THis helps admins to create a lease on the host while it is non reservable and host monitor not changing the status back to reservble - True
- 'disabled' field to use in host plugin and admins can create lease when host is non reservable
- - `openstack reservation host set <host id/name> --extra disabled=true`
- - `openstack reservation host unset <host id/name> --extra disabled`
- When disabled key is present in extra capabilities handle it by setting it or unsetting it based on the value for the key
- When disabled field is set to True for a host, the host also changes its reservable status to False.
- Disabled key can only be set by an admin
- - An admin can create a lease on a host that is non reservable
- - Use the policy.enforce to check if the blazar context is a an admin from keystone
- Add disabled flag alembic migration file
- notification callback to use set_reservable method
- - Fix tests that call matching hosts as users, so is_admin should return False
- - New tests for updating disabled field
- - new test for setting reservable True by monitor when disabled is True / False

AnishReddyRavula committed:
Add disabled field in ComputeHost model

This can be used by admins to specify that a host is disabled for maintainance and that marks that blazar monitor should not touch or alter this host's health while it is disabled (True). THis helps admins to create a lease on the host while it is non reservable and host monitor not changing the status back to reservble - True

AnishReddyRavula committed:
'disabled' field to use in host plugin and admins can create lease when host is non reservable

When disabled key is present in extra capabilities handle it by setting it or unsetting it based on the value for the key

When disabled field is set to True for a host, the host also changes its reservable status to False.

Disabled key can only be set by an admin

An admin can create a lease on a host that is non reservable

AnishReddyRavula committed:
Add disabled flag alembic migration file

AnishReddyRavula committed:
Use the policy.enforce to check if the blazar context is a an admin from keystone

AnishReddyRavula committed:
notification callback to use set_reservable method
Fix tests that call matching hosts as users, so is_admin should return False
New tests for updating disabled field
new test for setting reservable True by monitor when disabled is True / False

Mark-Powers commented on blazar/db/migration/alembic_migrations/versions/d5a379ff6ba3_add_disabled_in_compute_host.py:
```diff
@@ -0,0 +1,98 @@
+    op.alter_column('devices', 'device_driver',
+               existing_type=mysql.ENUM('zun', 'k3s'),
+               nullable=False)
+    op.drop_column('network_reservations', 'vfc_resources')
```
I think this got autogenerated for some reason? Can you remove it?

Mark-Powers commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -515,6 +520,24 @@ def delete_computehost(self, host_id):
+        # only admin can set/unset 'disabled' flag
+        if self._is_admin():
+            host_details = self.get_computehost(host_id)
+            new_disabled_flag = False if value is None else True
```
Should this not be `False if value is None else value`?

AnishReddyRavula replied:
The `value` is parsed as a string. As there is no schema or exact string the user can use (they can use 'true', 'True', 'T', or any string for that matter). The only thing matters is `set` or `unset` while the set needs a keykey-value pair, but unset needs only the key with value of `None`. So as long as `set` is called with some _value_ this will work. 

I am not sure about the proper way this can be documented.

Mark-Powers replied:
Can you share the commands you were using?

Mark-Powers commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -897,14 +932,16 @@ def notification_callback(self, event_type, payload):
 
     def set_reservable(self, resource, is_reservable):
+        if resource.get('disabled', False):
+            LOG.warn(f"{resource['hypervisor_hostname']} is disabled - cannot set reservable")
```
I would set this to `LOG.debug`, as it will print every time the monitor runs in most cases.

Mark-Powers reviewed (approved):
This change is great! Just a few minor comments. Did you test this via CLI? How exactly is that being used? I think right now the way the cli works you'd need `--extra disabled=true/false`

AnishReddyRavula committed:
Remove line in migration auto generated by alembic

addressing PR comments

change log warn to log debug if a host is disabled and admin is trying to set host reservable
